File: elixir/lib/day02.ex
defmodule Day02 do
  def get_input(file) do
    File.read!(file)
    |> String.split("\n", trim: true)
  end

  def compute([op, val], {depth, distance}) do
    val = String.to_integer(val)

    case op do
      "forward" -> {depth, distance + val}
      "down" -> {depth + val, distance}
      "up" -> {depth - val, distance}
    end
  end

  def compute([op, val], {depth, distance, aim}) do
    val = String.to_integer(val)

    case op do
      "forward" -> {depth + aim * val, distance + val, aim}
      "down" -> {depth, distance, aim + val}
      "up" -> {depth, distance, aim - val}
    end
  end

  def part1 do
    get_input("priv/day02")
    |> Enum.map(&String.split(&1, " ", trim: true))
    |> Enum.reduce({0, 0}, &compute/2)
    |> Tuple.product()
  end

  def part2 do
    get_input("priv/day02")
    |> Enum.map(&String.split(&1, " ", trim: true))
    |> Enum.reduce({0, 0, 0}, &compute/2)
    |> (fn {depth, distance, _} -> depth * distance end).()
  end
end

File: clojure/src/aoc/day02.clj
(ns aoc.day02
  (:require [aoc.core :as core]
            [clojure.string :as s]))

(defn normalize [s]
  (let [[dir v] (s/split s #" ")]
    [dir (Integer/parseInt v)]))

(defn calc-movement-p1 [[hor ver] [dir v]]
  (case dir
    "forward" [(+ hor v) ver]
    "up"      [hor (- ver v)]
    "down"    [hor (+ ver v)]))

(defn calc-movement-p2 [[hor ver aim] [dir v]]
  (case dir
    "forward" [(+ hor v) (+ ver (* v aim)) aim]
    "up"      [hor ver (- aim v)]
    "down"    [hor ver (+ aim v)]))

(defn part1 []
  (->> (core/get-input "day02")
       (map normalize)
       (reduce calc-movement-p1 [0 0])
       (reduce *)))

(defn part2 []
  (->> (core/get-input "day02")
       (map normalize)
       (reduce calc-movement-p2 [0 0 0])
       (take 2)
       (reduce *)))
